File: sponge-jruby/examples/script/rb/events_clone_policy.rb
# Sponge Knowledge Base
# Deep and shallow event clone policy.

java_import java.lang.Boolean
java_import java.lang.Integer
java_import java.util.Collections
java_import java.util.HashMap
java_import java.util.ArrayList

def onInit
    # Variables for assertions only
    $events = Collections.synchronizedMap(HashMap.new)
    $events.put("defaultClonePolicy", ArrayList.new)
    $events.put("deepClonePolicy", ArrayList.new)
    $events.put("shallowClonePolicy", ArrayList.new)
    $sponge.setVariable("events", $events)
end

class ClonePolicyTrigger < Trigger
    def onConfigure
        self.withEvents(["defaultClonePolicy", "deepClonePolicy", "shallowClonePolicy"])
    end
    def onRun(event)
        $events.get(event.name).add(event)
        self.logger.debug("Processing event: {}", event.name)
        map = event.get("map")
        self.logger.debug("map attribute (before) {}", map)
        map.put("a", "Value " + $events.get(event.name).size().to_s);
        self.logger.debug("map attribute (after) {}", map)
    end
end

def onStartup
    setEventAttributes = lambda { |event|
        hash = HashMap.new
        hash.put("a", "Value 0")
        hash.put("b", [Boolean::TRUE].to_java(:Boolean)) # Convert to Java array which is serializable.
        event.set("map", hash)
        event.set("integer", Integer.new(10))
    }

    $sponge.event("defaultClonePolicy").modify(setEventAttributes).sendAfter(100, 1000)
    $sponge.event("deepClonePolicy", EventClonePolicy::DEEP).modify(setEventAttributes).sendAfter(200, 1000)
    $sponge.event("shallowClonePolicy", EventClonePolicy::SHALLOW).modify(setEventAttributes).sendAfter(400, 1000)
end

File: sponge-kotlin/src/test/kotlin/examples/EventsClonePolicy.kt
/*
 * Sponge Knowledge Base
 * Deep and shallow event clone policy.
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.event.Event
import org.openksavi.sponge.event.EventClonePolicy
import org.openksavi.sponge.kotlin.KKnowledgeBase
import org.openksavi.sponge.kotlin.KTrigger
import java.util.Collections

class EventsClonePolicy : KKnowledgeBase() {
    var events = Collections.synchronizedMap(HashMap<String, MutableList<Event>>())

    override fun onInit() {
        // Variables for assertions only
        events.put("defaultClonePolicy", ArrayList())
        events.put("deepClonePolicy", ArrayList())
        events.put("shallowClonePolicy", ArrayList())
        sponge.setVariable("events", events)
    }

    class ClonePolicyTrigger : KTrigger() {
        override fun onConfigure() {
            withEvents("defaultClonePolicy", "deepClonePolicy", "shallowClonePolicy")
        }

        override fun onRun(event: Event) {
            var events: Map<String, MutableList<Event>> = sponge.getVariable("events")

            events.get(event.name)!!.add(event)
            logger.debug("Processing event: {}", event.name)
            var map: MutableMap<String, Any> = event.get("map")
            logger.debug("map attribute (before): {}", map)
            map.put("a", "Value " + events.get(event.name)!!.size)
            logger.debug("map attribute (after): {}", map)
        }
    }

    override fun onStartup() {
        fun setEventAttributes(event: Event) {
            val hash = HashMap<String, Any>()
            hash.put("a", "Value 0")
            hash.put("b", listOf(java.lang.Boolean.TRUE))
            event.set("map", hash)
            event.set("integer", Integer(10))
        }

        sponge.event("defaultClonePolicy").modify(::setEventAttributes).sendAfter(100, 1000)
        sponge.event("deepClonePolicy", EventClonePolicy.DEEP).modify(::setEventAttributes).sendAfter(200, 1000)
        sponge.event("shallowClonePolicy", EventClonePolicy.SHALLOW).modify(::setEventAttributes).sendAfter(400, 1000)
    }
}
